components/script/dom/notification.rs

```diff
@@ -0,0 +1,348 @@
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::bindings::trace::RootedTraceableBox;
+use crate::dom::bindings::utils::to_frozen_array;
+use crate::dom::eventtarget::EventTarget;
+use crate::dom::globalscope::GlobalScope;
+use crate::dom::permissions::get_descriptor_permission_state;
+use crate::dom::promise::Promise;
+use crate::dom::serviceworkerglobalscope::ServiceWorkerGlobalScope;
+use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
+
+/// <https://notifications.spec.whatwg.org/#api>
+#[dom_struct]
+pub struct Notification {
+    eventtarget: EventTarget,
+    title: DOMString,
+    body: DOMString,
+    #[ignore_malloc_size_of = "mozjs"]
+    data: Box<Heap<JSVal>>,
+    dir: NotificationDirection,
+    image: Option<USVString>,
+    icon: Option<USVString>,
+    badge: Option<USVString>,
+    lang: DOMString,
+    silent: Option<bool>,
+    tag: DOMString,
+    #[no_trace] // ImmutableOrigin is not traceable
+    origin: ImmutableOrigin,
+    // TODO: vibrate not implemented yet
+    // vibrate: Option<UnionTypes::UnsignedLongOrUnsignedLongSequence>,
+    timestamp: u64,
+    renotify: bool,
+    require_interaction: bool,
+    #[ignore_malloc_size_of = "NotificationAction"] // malloc not implement for NotificationAction
+    actions: Vec<NotificationAction>,
+    closed: Cell<bool>,
+}
+
+impl Notification {
+    #[allow(crown::unrooted_must_root, clippy::too_many_arguments)]
+    pub fn new(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+        can_gc: CanGc,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+        origin: Option<ImmutableOrigin>,
+        base_url: Option<ServoUrl>,
+        fallback_timestamp: u64,
+    ) -> Fallible<DomRoot<Self>> {
+        let notification = Notification::new_inherited(
+            global,
+            title,
+            options,
+            origin,
+            base_url,
+            fallback_timestamp,
+        )?;
+        Ok(reflect_dom_object_with_proto(
+            Box::new(notification),
+            global,
+            proto,
+            can_gc,
+        ))
+    }
+
+    /// <https://notifications.spec.whatwg.org/#create-a-notification-with-a-settings-object>
+    fn new_inherited(
+        global: &GlobalScope,
+        title: DOMString,
+        options: RootedTraceableBox<NotificationOptions>,
+        origin: Option<ImmutableOrigin>,
+        base_url: Option<ServoUrl>,
+        fallback_timestamp: u64,
+    ) -> Fallible<Self> {
+        if options.silent.is_some() && options.vibrate.is_some() {
+            return Err(Error::Type(
+                "silent and vibrate can not be set at the same time".to_string(),
+            ));
+        }
+        if options.renotify && options.tag.is_empty() {
+            return Err(Error::Type(
+                "tag must be set if renotify is true".to_string(),
+            ));
+        }
```

Comment:
Are these restrictions from the specification? If so please link to the specification text that says what kind of error to return or otherwise explain why you are throwing here.